osu.Server.Spectator/Startup.cs

```diff
@@ -80,8 +86,21 @@ protected virtual void ConfigureAuthentication(IServiceCollection services)
         }
 
         // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
-        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+        public void Configure(IApplicationBuilder app, IWebHostEnvironment env, IHostApplicationLifetime lifetime)
         {
+            var shutdownManager = app.ApplicationServices.GetService<GracefulShutdownManager>();
+
+            Debug.Assert(shutdownManager != null);
+
+            // This is done here, but it happens too late in shutdown.
+            // The more important registration is in EntityStore.
+            lifetime.ApplicationStopping.Register(shutdownManager.WaitForSafeShutdown);
+
+            // Importantly, we don't care to block `MultiplayerClientState` because they can only be created if a `ServerMultiplayerRoom` is first in existence.
+            // More so, we want to allow these states to be created so existing rooms can continue to function until they are disbanded.
+            shutdownManager.AddDependentStore(app.ApplicationServices.GetRequiredService<EntityStore<ServerMultiplayerRoom>>());
+            shutdownManager.AddDependentStore(app.ApplicationServices.GetRequiredService<EntityStore<SpectatorClientState>>());
```

Comment:
What about `EntityStore<MultiplayerClientState>`? Given 1 `SpectatorClientState` is basically the same as a `MultiplayerClientState`, I don't see an issue with having it, and _not_ having it would still new allow clients to connect.

Also, technically you can do this inside the `GracefulShutdownManager` ctor too.

Reply:
I'm 99.9% sure there's an inline comment mentioning this.

Adding it will cause new clients to be *unable to join existing rooms*.

Reply:
How is this going to be routed? There's going to be two servers up simultaneously, and I don't think we have any way to identify the build ID via the connection, right?

So within a 6 hour timespan, _every new user_ that connects, regardless of being on a new build or not, will not be able to create rooms or spectate, or have their replays saved server-side?

I thought we were going down a different path of preserving existing connections and moving new connections to the new server.

Reply:
This PR makes it so:

- New rooms cannot be created
- New spectator sessions cannot be started

It leaves in place:

- Existing rooms can continue to be played
- Users still on the server can still connect to existing rooms

I touched on the next steps in the OP here:

> Eventually we may want to use multiple endpoints or sticky cookies/headers to persist server across connection loss (aka cloudflare support, since cloudflare arbitrarily drops websocket connections) but for now we are avoiding this by dropping cloudflare temporarily.

And the inline comment here:

https://github.com/ppy/osu-server-spectator/pull/133/files#diff-90658690a1d84ef42d78f75a985211e9140da7fdadbcb3b5562eee9c5d894fd7R171-R174

Please let me know how I could have worded this more clearly or if something is missing which caused your confusion.

Reply:
> So within a 6 hour timespan, _every new user_ that connects, regardless of being on a new build or not, will not be able to create rooms or spectate, or have their replays saved server-side?

New users connect to a new server. No new connections can reach a terminating server, only already connected ones.

Reply:
So based on that quote what you're trying to say is that the intention is if someone has a power outage then they're screwed, but if their connection drops momentarily then they can join back in to the old server?

And it's an intention because we don't actually have support for this yet.

It just seems weird and overly complex to me. I'm going to guess that your reasoning for this is that players on the server should still be able to join rooms so as to not shaft tournaments, however what about players that don't even have their game started up? I guess they're just shafted anyway?

Reply:
Connection dropping momentarily *cannot reconnect to the old server*. Did you want to voice? There should be nothing complicated about it.

Also there is no "shafting" involved. We already discussed having a window where tournaments would know not to make rooms. This does not resolve that issue. The only thing it does is allow initial deployment of automatic release handling. It's equivalent to what I've been doing manually but with some new safeties.

Reply:
See discussions later on in the thread.